File: 245-shortest-word-distance-iii.cpp
class Solution {
public:
    int shortestWordDistance(vector<string>& wordsDict, string word1, string word2) {
        vector<int> w1;
        vector<int> w2;
        for (int i = 0; i < wordsDict.size(); i++) {
            if (wordsDict[i] == word1) {
                w1.push_back(i);
            }
            if (wordsDict[i] == word2) {
                w2.push_back(i);
            }
        }
        int minn = INT_MAX;
        int i = 0; 
        int j = 0;
        while (i < w1.size() && j < w2.size()) {
            if (w1[i] < w2[j]) {
                minn = min(minn, w2[j] - w1[i]);
                i++;
            } else if (w2[j] < w1[i]) {
                minn = min(minn, w1[i] - w2[j]);
                j++;
            } else {
                i++;
            }
        }
        return minn;
    }
};

File: 245-shortest-word-distance-iii.py
class Solution:
    def shortestWordDistance(self, wordsDict: List[str], word1: str, word2: str) -> int:
        if word1 == word2:
            last = -1
            minn = math.inf
            for i in range(len(wordsDict)):
                word = wordsDict[i]
                if word == word1:
                    if last == -1:
                        last = i
                        continue
                    minn = min(minn, i - last)
                    last = i
            return minn
        last1 = -1
        last2 = -1
        minn = math.inf
        for i in range(len(wordsDict)):
            word = wordsDict[i]
            if word == word1:
                if last2 == -1:
                    last1 = i
                    continue
                minn = min(minn, i - last2)
                last1 = i
            elif word == word2:
                if last1 == -1:
                    last2 = i
                    continue
                minn = min(minn, i - last1)
                last2 = i
        return minn